pl_bolts/datamodules/sklearn_datamodule.py

```diff
@@ -1,280 +1,177 @@
         self.pin_memory = pin_memory
         self.drop_last = drop_last
+        self.persistent_workers = persistent_workers and num_workers > 0
 
-        # shuffle x and y
-        if shuffle and _SKLEARN_AVAILABLE:
-            X, y = sk_shuffle(X, y, random_state=random_state)
-        elif shuffle and not _SKLEARN_AVAILABLE:  # pragma: no cover
-            raise ModuleNotFoundError(
-                "You want to use shuffle function from `scikit-learn` which is not installed yet."
+        if isinstance(self._test_dataset, ArrayDataset):
+            self.x_train, self.x_val, self.y_train, self.y_val = self._sklearn_train_test_split(
+                self.data, self.target, split=self.val_size
+            )
+        else:
+            _x, self.x_test, _y, self.y_test = self._sklearn_train_test_split(
+                self.data, self.target, split=self.test_size
+            )
+            self.x_train, self.x_val, self.y_train, self.y_val = self._sklearn_train_test_split(
+                _x, _y, split=self.val_size
             )
 
-        val_split = 0 if x_val is not None or y_val is not None else val_split
-        test_split = 0 if x_test is not None or y_test is not None else test_split
+        self.train_dataset = ArrayDataset(self.x_train, self.y_train)
+        self.val_dataset = ArrayDataset(self.x_val, self.y_val)
+        self.test_dataset = (
+            ArrayDataset(self.x_test, self.y_test)
+            if self.x_test is not None and self.y_test is not None
+            else self._test_dataset
+        )
 
-        hold_out_split = val_split + test_split
-        if hold_out_split > 0:
-            val_split = val_split / hold_out_split
-            hold_out_size = math.floor(len(X) * hold_out_split)
-            x_holdout, y_holdout = X[:hold_out_size], y[:hold_out_size]
-            test_i_start = int(val_split * hold_out_size)
-            x_val_hold_out, y_val_holdout = x_holdout[:test_i_start], y_holdout[:test_i_start]
-            x_test_hold_out, y_test_holdout = x_holdout[test_i_start:], y_holdout[test_i_start:]
-            X, y = X[hold_out_size:], y[hold_out_size:]
+    def train_dataloader(self) -> DataLoader:
+        return self._data_loader(self.train_dataset, shuffle=True)
 
-        # if don't have x_val and y_val create split from X
-        if x_val is None and y_val is None and val_split > 0:
-            x_val, y_val = x_val_hold_out, y_val_holdout
+    def val_dataloader(self) -> DataLoader:
+        return self._data_loader(self.val_dataset)
 
-        # if don't have x_test, y_test create split from X
-        if x_test is None and y_test is None and test_split > 0:
-            x_test, y_test = x_test_hold_out, y_test_holdout
+    def test_dataloader(self) -> DataLoader:
+        return self._data_loader(self.test_dataset)
 
-        self._init_datasets(X, y, x_val, y_val, x_test, y_test)
+    def _sklearn_train_test_split(self, x, y, split: Optional[Union[float, int]] = None):
```

Comment:
It might make sense to have either an error from us saying sklearn is needed. On the other hand, this is such a small amount of code, that it might also make sense to provide our own implementation

Reply:
I would be happy to move away from sklearn as only one function is used. Using `random_split` from torch is a potential option. I will have a think.